File: hphp/hack/src/hhbc/hhas_property.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

type t = {
  property_attributes: Hhas_attribute.t list;
  property_visibility: Aast.visibility;
  property_is_abstract: bool;
  property_is_static: bool;
  property_is_deep_init: bool;
  property_is_const: bool;
  property_is_lsb: bool;
  property_is_no_bad_redeclare: bool;
  property_has_system_initial: bool;
  property_no_implicit_null: bool;
  property_initial_satisfies_tc: bool;
  property_is_late_init: bool;
  property_name: Hhbc_id.Prop.t;
  property_initial_value: Typed_value.t option;
  property_initializer_instrs: Instruction_sequence.t option;
  property_type_info: Hhas_type_info.t;
  property_doc_comment: string option;
}

let make
    property_attributes
    property_visibility
    property_is_abstract
    property_is_static
    property_is_deep_init
    property_is_const
    property_is_lsb
    property_is_no_bad_redeclare
    property_has_system_initial
    property_no_implicit_null
    property_initial_satisfies_tc
    property_is_late_init
    property_name
    property_initial_value
    property_initializer_instrs
    property_type_info
    property_doc_comment =
  {
    property_attributes;
    property_visibility;
    property_is_abstract;
    property_is_static;
    property_is_deep_init;
    property_is_const;
    property_is_lsb;
    property_is_no_bad_redeclare;
    property_has_system_initial;
    property_no_implicit_null;
    property_initial_satisfies_tc;
    property_is_late_init;
    property_name;
    property_initial_value;
    property_initializer_instrs;
    property_type_info;
    property_doc_comment;
  }

let name hhas_property = hhas_property.property_name

let attributes hhas_property = hhas_property.property_attributes

let is_private hhas_property = hhas_property.property_visibility = Aast.Private

let is_protected hhas_property =
  hhas_property.property_visibility = Aast.Protected

let is_public hhas_property = hhas_property.property_visibility = Aast.Public

let visibility hhas_property = hhas_property.property_visibility

let is_abstract hhas_property = hhas_property.property_is_abstract

let is_static hhas_property = hhas_property.property_is_static

let is_deep_init hhas_property = hhas_property.property_is_deep_init

let initial_value hhas_property = hhas_property.property_initial_value

let initializer_instrs hhas_property = hhas_property.property_initializer_instrs

let is_const hhas_property = hhas_property.property_is_const

let is_lsb hhas_property = hhas_property.property_is_lsb

let is_no_bad_redeclare hhas_property =
  hhas_property.property_is_no_bad_redeclare

let has_system_initial hhas_property = hhas_property.property_has_system_initial

let no_implicit_null hhas_property = hhas_property.property_no_implicit_null

let initial_satisfies_tc hhas_property =
  hhas_property.property_initial_satisfies_tc

let is_late_init hhas_property = hhas_property.property_is_late_init

let type_info hhas_property = hhas_property.property_type_info

let doc_comment hhas_property = hhas_property.property_doc_comment

File: hphp/hack/src/hhbc/hhas_property.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use hhas_attribute_rust::HhasAttribute;
use hhbc_id_rust as hhbc_id;
use instruction_sequence::InstrSeq;
use oxidized::{aast_defs::Visibility, doc_comment::DocComment};

extern crate bitflags;
use bitflags::bitflags;

bitflags! {
    pub struct HhasPropertyFlags: u16 {
        const IS_ABSTRACT = 1 << 0;
        const IS_STATIC = 1 << 1;
        const IS_DEEP_INIT = 1 << 2;
        const IS_CONST = 1 << 3;
        const IS_LSB = 1 << 4;
        const IS_NO_BAD_REDECLARE = 1 << 5;
        const HAS_SYSTEM_INITIAL = 1 << 6;
        const NO_IMPLICIT_NULL = 1 << 7;
        const INITIAL_SATISFIES_TC = 1 << 8;
        const IS_LATE_INIT = 1 << 9;
        const IS_READONLY = 1 << 10;
    }
}

#[derive(Debug)]
pub struct HhasProperty<'id> {
    pub name: hhbc_id::prop::Type<'id>,
    pub flags: HhasPropertyFlags,
    pub attributes: Vec<HhasAttribute>,
    pub visibility: Visibility,
    pub initial_value: Option<runtime::TypedValue>,
    pub initializer_instrs: Option<InstrSeq>,
    pub type_info: hhas_type::Info,
    pub doc_comment: Option<DocComment>,
}
impl<'a> HhasProperty<'a> {
    pub fn is_private(&self) -> bool {
        self.visibility == Visibility::Private
    }
    pub fn is_protected(&self) -> bool {
        self.visibility == Visibility::Protected
    }
    pub fn is_public(&self) -> bool {
        self.visibility == Visibility::Public
    }
    pub fn is_late_init(&self) -> bool {
        self.flags.contains(HhasPropertyFlags::IS_LATE_INIT)
    }
    pub fn is_no_bad_redeclare(&self) -> bool {
        self.flags.contains(HhasPropertyFlags::IS_NO_BAD_REDECLARE)
    }
    pub fn initial_satisfies_tc(&self) -> bool {
        self.flags.contains(HhasPropertyFlags::INITIAL_SATISFIES_TC)
    }
    pub fn no_implicit_null(&self) -> bool {
        self.flags.contains(HhasPropertyFlags::NO_IMPLICIT_NULL)
    }
    pub fn has_system_initial(&self) -> bool {
        self.flags.contains(HhasPropertyFlags::HAS_SYSTEM_INITIAL)
    }
    pub fn is_const(&self) -> bool {
        self.flags.contains(HhasPropertyFlags::IS_CONST)
    }
    pub fn is_deep_init(&self) -> bool {
        self.flags.contains(HhasPropertyFlags::IS_DEEP_INIT)
    }
    pub fn is_lsb(&self) -> bool {
        self.flags.contains(HhasPropertyFlags::IS_LSB)
    }
    pub fn is_static(&self) -> bool {
        self.flags.contains(HhasPropertyFlags::IS_STATIC)
    }
    pub fn is_readonly(&self) -> bool {
        self.flags.contains(HhasPropertyFlags::IS_READONLY)
    }
}
